### PR TITLE
[Serializer] Adding `ignore_invalid_datetimes` flag to Serialization Context to avoid exceptions

### PR DESCRIPTION
**Pull request by liorchamla**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31596
| License       | MIT
| Doc PR        | TODO

I came to the idea of allowing invalid date strings to be denormalize _without throwing an exception_ because I often want to return a well formated violations list with real error messages to the end user/developer.

Actually, when someone submits an invalid date format, instead of receiving a cool violation message generated by the Validator Component (for instance), he receive an exception because the DateTimeNormalizer::denormalize method can't instanciate a DateTimeInterface instance.

This is a cool and normal behavior but we must allow API developers to choose how they want to manage this type of error and to be able to let it pass until Validation process handles it.

PS : I'm really sorry if I did something wrong here, this is my first PR and I would love to hear from you on what I can do next time to give it a better shape !

### CODE REVIEW

**Comment by liorchamla** [comment]
By the way, I changed the CHANGELOG.md in the Serializer folder but I don't know if it is what you wanted from me, let me know if I was wrong.

One other thing, Travis CI failed the tests but I looked into it and the failure occurs in the HttpKernel Component, so how do we deal with that ?

File: src/Symfony/Component/Serializer/CHANGELOG.md
Comment (dunglas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]:
```suggestion
4.4.0
```

File: src/Symfony/Component/Serializer/CHANGELOG.md
Comment (dunglas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: This name could be confusing because this option doesn't really made the serializer "ignoring" the passed value, it keeps it. What do you think about changing to `datetime_allow_invalid`?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]:
```suggestion
            if ($context[self::IGNORE_INVALID_DATETIMES] ?? $this->defaultContext[self::IGNORE_INVALID_DATETIMES]) {
```

You must also set `$this->defaultContext[self::IGNORE_INVALID_DATETIMES]` to `false` in the constructor.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/CHANGELOG.md
Comment (liorchamla) [reply src/Symfony/Component/Serializer/CHANGELOG.md]: Change was made (and tested)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: This change was also made on the PR :-)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]:
```suggestion
    public const ALLOW_INVALID = 'datetime_allow_invalid';
```

For consistency with the options defined above.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: I must agree that having a DATETIME_ALLOW_INVALID setted in the DateTimeNormalizer is a bit redundant :-)

Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Change has been made right away ! ⚔️

**Review by dunglas — APPROVED** [review]

**Comment by liorchamla** [comment]
Yeah ! Thx so much @dunglas for all your help on this ! So proud to be a tiny little part of the project <3

**Review by maxhelias — APPROVED** [review]

**Comment by dunglas** [comment]
No, don’t close it. We have to wait for the approval of another core team member now.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: proposal:
When setting this option to true, denormalization will ignore invalid dates and leave them as strings. This can be useful to replace exceptions thrown by the serializer by more contextual exceptions thrown by a later validation step.

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: this should be reverted, it's unrelated

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (Simperfit) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]:
```suggestion
     * Setting this tag to true will allow you to retrieve the real instance of DateTime
```

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (Simperfit) [inline src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: Can we do that ? is it needed ?

Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: But actually, this is a problem: this type is the only legal return type, and we could actually enforce it in master with a real return type.
This means this PR will break. Is this a hint this  is not the correct approach?
The denormalization logic should throw an exception to me, but maybe we would add a generic option that would allow the caller to catch the exception and leave the original value in place?

**Comment by fabpot** [comment]
I'm skeptical about this change. Some questions to better reason about it: Is `Datetime` special? If not, why not do the same for other normalizer? But then, it's like adding an option to bypass the normalization process, which does not seem something we want to do. It looks like the wrong approach to me.

**Comment by liorchamla** [comment]
I dont know the whole component as you all do fellows, I will just comply to whatever your opinion is.

I understand @fabpot and the approach he suggests : when using the serializer by yourself, just put it in a try / catch and see if you run an exception, just deal with it (note that, I think, PHP does not have a specific exception class for DateTime::__construct() failure, so you can't grab it easily to address it).

But for instance, if you use ApiPlatform, at what time do you have control over serialization process other than the glorious $context options ? I'm sure you will know how to answer to that but as a simple consumer of the component, this is really painfull not to simply retrieve my Validations violations and having to deal with DateTime::__construct Exception.

This is why it seemed legit to me to get it official with a flag. At first I came with the idea that it could be simply done when activating the 'disable_type_enforcement' flag. Which I usualy use to take over the control of my violations messages.

No worries if you think this is not good :) I would gladly hear what other approach can be made here and which one would comply the best to our (end users of the component, noobs in 90% of cases who just want it to work :D) practices.

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: I don't know if it is **needed**, but I was thinking that if the function will return at some point a string (when the DateTimeNormalizer fails to instanciate a DateTime object and that the flag datetime_allow_invalid is set to true), we normaly have to signal it in the PHPDoc no ?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: You mean, create a new Exception class which will be more specific than the NotNormalizableValueException ?

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: You mean, create a new Exception class which will be more specific than the NotNormalizableValueException ?

**Comment by nicolas-grekas** [comment]
In my comment above I was suggesting a solution:

> maybe we should add a generic option that would allow the caller to catch the exception and leave the original value in place?

This maps to the comment from @fabpot "why not do the same for other normalizer?", and to yours @liorchamla "it could be simply done when activating the 'disable_type_enforcement",

so can't we do it via `disable_type_enforcement`, in `AbstractObjectNormalizer`?

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: Or catch `NotNormalizableValueException` and leave original values in place when `disable_type_enforcement` is set (or another option if the semantics doesn't fit?)

**Comment by liorchamla** [comment]
> I'm skeptical about this change. Some questions to better reason about it: Is `Datetime` special? If not, why not do the same for other normalizer? But then, it's like adding an option to bypass the normalization process, which does not seem something we want to do. It looks like the wrong approach to me.

This is why at first, I wanted to rely on `disable_type_enforcement` flag to handle this :-)

If DateTime is not special, then why can't we achive the same thing that we can when someone sends "bonjour" in a property that should be an int ?

With disable_type_enforcement we can handle typings problems with the validation process, on every value, but not on a date. This is not normal to me as a simple developer.

We should work around this : we can throw a nice violation list with human messages on each property of an object, but not datetimes (which continues to throw us a simple DateTime::__construct() exception ôÔ).

**Comment by liorchamla** [comment]
> so can't we do it via `disable_type_enforcement`, in `AbstractObjectNormalizer`?

That was my first approach and first PR about this subject but I closed it after a discussion with @dunglas :-)

**Comment by nicolas-grekas** [comment]
I didn't review the original approach so I can't comment on its merits or drawbacks :)

**Comment by liorchamla** [comment]
> I didn't review the original approach so I can't comment on its merits or drawbacks :) (but the current approach is no-go for the given reasons personnaly)

No problem for the current approach, I understand all the problems it may cause. But @nicolas-grekas, if you also think `disable_type_enforcement` is the right way to go, then we agree ! Great :heart: 

We could simply change 2 lines of this PR to fallback on this flag instead of `datetime_allow_invalid` :-)

**Comment by nicolas-grekas** [comment]
Maybe we can close this PR and open a new one with the other approach? (to keep some history)

**Comment by liorchamla** [comment]
> Maybe we can close this PR and open a new one with the other approach? (to keep some history)
I can also make a new commit here, it really is the same approach, except we replace the new tag `datetime_allow_invalid` with the old good `disable_type_enforcement`

I don't know if this is what you had in mind when talking about an **other** approach, involving catching exceptions on a higher level :x

**Comment by nicolas-grekas** [comment]
I meant using `disable_type_enforcement` in `AbstractObjectNormalizer` (not in `DateTimeNormalizer`)

**Comment by liorchamla** [comment]
> I meant using `disable_type_enforcement` in `AbstractObjectNormalizer` (not in `DateTimeNormalizer`)

Since I don't know exactly how you imagine that, and since `AbstractObjectNormalizer` is way more complicated than the `DateTimeNormalizer`, I don't know if I would be able to do what you want :x

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php
Comment (liorchamla) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: By the way, in a world where this PR would have been approved, your proposal for this comment is much better than mine ^^

File: src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php
Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php]: Not related, but maybe that we should allow this method to return mixed. There are valid use cases where you want to “denormalize” a string as another string.
I’ll try the changes introduced in master with API Platform, but I think they’ll break somebody of our normalizers.

**Comment by dunglas** [comment]
Actually, the best fix would probably be to finish the patch proposed in  https://github.com/symfony/symfony/pull/27136.

It would allow to catch all type errors in a user friendly way. It would also pave the way to allow to correct errors but let the normalizers populating invalid values as suggested by Nicolas.

**Comment by liorchamla** [comment]
> Actually, the best fix would probably be to finish the patch proposed in #27136.

I will test this PR in order to see at what point it answers to the same proposal I made :-) just by reading I can't see if it aims the same goal.

**Comment by lyrixx** [comment]
This PR also fix #27824 isn't ?

**Comment by liorchamla** [comment]
> This PR also fix #27824 isn't ?

Yes this was the exact goal of this PR but it seems it is not the way reviewers would like to go. 

I need time to test an other approach given by an other PR quoted @dunglas :

> Actually, the best fix would probably be to finish the patch proposed in #27136.
> It would allow to catch all type errors in a user friendly way. It would also pave the way to allow to collect errors but let the normalizers populating invalid values as suggested by Nicolas.

**Comment by fabpot** [comment]
Closing as this PR is stale and the approach won't be merged as is.

**Pull request closed without merging**
